### PR TITLE
gpu training with parquet reader

### PR DESCRIPTION
**Pull request by imatiach-msft**
Adding GPU training with Parquet reader

### CODE REVIEW

File: src/cntk-train/src/main/scala/CommandBuilders.scala [hunk 1]
```diff
@@ -181,8 +184,9 @@ class MPICommandBuilder(log: Logger,
         Seq("scp", "-i", identity, "-r", "-o", "StrictHostKeyChecking=no",
           fileInputPath, s"$gpuUser:$fileInputPath"))
     }
+    val runMPI = "''" + s"$exportClasspath $command $mpiArgs $cntkArgs" + "''"
     // Run the mpi command
-    printOutput(Seq("ssh", "-i", identity, gpuUser, command, mpiArgs, cntkArgs, "parallelTrain=true"))
+    printOutput(Seq("ssh", "-i", identity, gpuUser, runMPI))
```
Comment (tongwen11) [inline src/cntk-train/src/main/scala/CommandBuilders.scala]: Just want to make sure we are not losing "parallelTrain=true"

File: src/cntk-train/src/main/scala/DataConversion.scala [hunk 2]
```diff
@@ -106,6 +103,74 @@ object DataTransferUtils {
...
+      }
+    // Note: order matters, features column needs to come first
+    val cols = data.select(feats, label).columns.flatMap(name => {
+      if ((name == label && labelForm == CNTKLearner.denseForm) ||
+        (name == feats && featureForm == CNTKLearner.denseForm)) {
+        Seq(toDenseArray(data.col(name)).alias(name))
+      } else {
+        Seq(getNZs(data.col(name)).alias(name + "_size"),
+          getIdxes(data.col(name)).alias(name + "_indices"),
+          getVals(data.col(name)).alias(name) + "_values")
```
Comment (tongwen11) [inline src/cntk-train/src/main/scala/DataConversion.scala]: Are we missing a closing ")" to match "Seq("

Comment (imatiach-msft) [reply src/cntk-train/src/main/scala/DataConversion.scala]: good catch, the string:
+ "_values"
needs to be inside alias() I believe

**Comment by tongwen11** [comment]
Ilya,

Do you want to make add the top-level try-catch for the final cleanup in this check-in or a separate one?

Another touch up item is to handle cases where the DF passed to CNTKLearner has columns which are not Vector type.

File: src/cntk-train/src/main/scala/CommandBuilders.scala (hunk 1 above)
Comment (imatiach-msft) [reply src/cntk-train/src/main/scala/CommandBuilders.scala]: it didn't seem to make a difference for me, I believe it's already in the override config file

**Comment by imatiach-msft** [comment]
I can make the top level try catch in a separate checkin

File: src/cntk-train/src/main/scala/BrainscriptBuilder.scala [hunk 3]
```diff
@@ -74,13 +88,39 @@ class BrainScriptBuilder {
...
+  }
+
+  def parquetReaderConfig(loc: String, ipstring: String): String = {
+    s"""    reader = {
+      |        deserializers = (
+      |                        [
+      |                            type = "DataFrameDeserializer"
+      |                            module = "Cntk.Deserializers.DF"
+      |                            $ipstring
+      |                            hdfs = {
```
Comment (drdarshan) [inline src/cntk-train/src/main/scala/BrainscriptBuilder.scala]: Nice! is the port number fixed or does it need to be made a parameter.

File: src/cntk-train/src/main/scala/CNTKLearner.scala [hunk 4]
```diff
@@ -16,12 +16,10 @@ import org.apache.spark.sql._
...
 
 trait CNTKParams extends MMLParams {
-
-  // This is only needed until Train* accepts CNTKLearner instead of CL acting like Train*
-  val labelsColumnName = StringParam(this, "labelsColumnName", "Label col", "labels")
+  val labelsColumnName = StringParam(this, "labelsColumnName", "label column name", "labels")
   def getLabelsColumnName: String = $(labelsColumnName)
 
-  val featuresColumnName = StringParam(this, "featuresColumnName", "feats col", "features")
+  val featuresColumnName = StringParam(this, "featuresColumnName", "features column name", "features")
```
Comment (drdarshan) [inline src/cntk-train/src/main/scala/CNTKLearner.scala]: Sentence case for docs please :-)

File: src/cntk-train/src/main/scala/CommandBuilders.scala [hunk 5]
```diff
@@ -161,14 +165,13 @@ class MPICommandBuilder(log: Logger,
 
     // Run commands below only if writing input data to HDFS
     var fileDirStr: String = ""
-    if (hdfsPath.isDefined) {
+    if (hdfsPath.isDefined && dataFormat == CNTKLearner.textDataFormat) {
```
Comment (drdarshan) [inline src/cntk-train/src/main/scala/CommandBuilders.scala]: I think the HDFS mounting is orthogonal to text format vs. Parquet. At some point, we can add libhdfs support to the CNTK text reader for example.

File: src/cntk-train/src/main/scala/BrainscriptBuilder.scala (hunk 3 above)
Comment (imatiach-msft) [reply src/cntk-train/src/main/scala/BrainscriptBuilder.scala]: it needs to be fixed according to Tong

File: src/cntk-train/src/main/scala/CNTKLearner.scala (hunk 4 above)
Comment (imatiach-msft) [reply src/cntk-train/src/main/scala/CNTKLearner.scala]: done, Eli has already modified the other params

File: src/cntk-train/src/main/scala/BrainscriptBuilder.scala (hunk 3 above)
Comment (tongwen11) [reply src/cntk-train/src/main/scala/BrainscriptBuilder.scala]: We can create a constant like CNTKLearner.rpcPortNumber to represent RPC protocol port number and use it to replace the port number in the template.

File: src/cntk-train/src/main/scala/CommandBuilders.scala [hunk 6]
```diff
@@ -114,10 +120,8 @@ class MPICommandBuilder(log: Logger,
...
   def runCommand(): Unit = {
-    val pathEnvVar = "PATH=/usr/bin/cntk/cntk/bin:$PATH"
-    val ldLibraryPathEnvVar =
-      "LD_LIBRARY_PATH=/usr/bin/cntk/cntk/lib:/usr/bin/cntk/cntk/dependencies/lib:$LD_LIBRARY_PATH"
-    val mpiArgs = s" -x $pathEnvVar -x $ldLibraryPathEnvVar --npernode ${nodeConfig.head._2} "
+    val exportClasspath = "export CLASSPATH=$($HADOOP_HOME/bin/hadoop classpath --glob); "
+    val mpiArgs = s" --npernode ${nodeConfig.head._2} "
     val cntkArgs = "cntk " + configs
       .map(c => if (fileBased) s"configFile=${configToFile(c)} " else c.text.mkString(" "))
       .mkString(" ")
```
Comment (drdarshan) [inline src/cntk-train/src/main/scala/CommandBuilders.scala]: Please hoist the location of the SSH key pair into a constant somewhere. We might want to make it configurable.

File: src/cntk-train/src/main/scala/BrainscriptBuilder.scala (hunk 3 above)
Comment (imatiach-msft) [reply src/cntk-train/src/main/scala/BrainscriptBuilder.scala]: done

File: src/cntk-train/src/main/scala/CommandBuilders.scala (hunk 6 above)
Comment (imatiach-msft) [reply src/cntk-train/src/main/scala/CommandBuilders.scala]: done

**Comment by imatiach-msft** [comment]
I added the top level try/finally based on Tong's recommendation and updated the code to use several constants

**Pull request merged**